PRINCIPAL COMPONENT ANALYSIS

In [ ]:
#IMPORT LIBRARIES
from sklearn import datasets
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB,BernoulliNB, MultinomialNB
from sklearn.linear_model import LogisticRegression

import pandas as pd

#LOAD BREAST CANCER DATASET
breast = datasets.load_breast_cancer()


In [ ]:
#CHECK DATA PROPERTIES AND FUNCTIONS
dir(breast)

['DESCR',
 'data',
 'data_module',
 'feature_names',
 'filename',
 'frame',
 'target',
 'target_names']

In [ ]:
#VIEW FEW SAMPLE DATA
breast.data[0:2]

array([[1.799e+01, 1.038e+01, 1.228e+02, 1.001e+03, 1.184e-01, 2.776e-01,
        3.001e-01, 1.471e-01, 2.419e-01, 7.871e-02, 1.095e+00, 9.053e-01,
        8.589e+00, 1.534e+02, 6.399e-03, 4.904e-02, 5.373e-02, 1.587e-02,
        3.003e-02, 6.193e-03, 2.538e+01, 1.733e+01, 1.846e+02, 2.019e+03,
        1.622e-01, 6.656e-01, 7.119e-01, 2.654e-01, 4.601e-01, 1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, 1.326e+03, 8.474e-02, 7.864e-02,
        8.690e-02, 7.017e-02, 1.812e-01, 5.667e-02, 5.435e-01, 7.339e-01,
        3.398e+00, 7.408e+01, 5.225e-03, 1.308e-02, 1.860e-02, 1.340e-02,
        1.389e-02, 3.532e-03, 2.499e+01, 2.341e+01, 1.588e+02, 1.956e+03,
        1.238e-01, 1.866e-01, 2.416e-01, 1.860e-01, 2.750e-01, 8.902e-02]])

In [ ]:
#VIEW FEATURE NAMES
breast.feature_names

array(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error',
       'fractal dimension error', 'worst radius', 'worst texture',
       'worst perimeter', 'worst area', 'worst smoothness',
       'worst compactness', 'worst concavity', 'worst concave points',
       'worst symmetry', 'worst fractal dimension'], dtype='<U23')

In [ ]:
#VIEW TARGET NAMES
breast.target_names

array(['malignant', 'benign'], dtype='<U9')

In [ ]:
#VIEW FEW ENCODED NUMERICAL TARGETS
breast.target[0:3]

array([0, 0, 0])

In [ ]:
#FORM A PANDAS DATAFRAME WITH BREAST CANCER DATA WITH INDEPENDENT FEATURE VALUES
df = pd.DataFrame(breast.data,columns=breast.feature_names)
df.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   

   worst area  worst smoothness  worst compactness  worst concavity  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   worst concave points  worst symmetry  worst fractal dimension  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 30 columns]

In [ ]:
#ADD DEPENDANT VARIABLE COLUMN (TARGET)
df['target'] = breast.target
df[50:70]

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
50       11.760         21.60           74.72      427.9          0.08637   
51       13.640         16.34           87.21      571.8          0.07685   
52       11.940         18.24           75.71      437.6          0.08261   
53       18.220         18.70          120.30     1033.0          0.11480   
54       15.100         22.02           97.26      712.8          0.09056   
55       11.520         18.75           73.34      409.0          0.09524   
56       19.210         18.57          125.50     1152.0          0.10530   
57       14.710         21.59           95.55      656.9          0.11370   
58       13.050         19.31           82.61      527.2          0.08060   
59        8.618         11.79           54.34      224.5          0.09752   
60       10.170         14.88           64.55      311.9          0.11340   
61        8.598         20.98           54.66      221.8          0.12430   
62       14.250         22.15           96.42      645.7          0.10490   
63        9.173         13.86           59.20      260.9          0.07721   
64       12.680         23.84           82.69      499.0          0.11220   
65       14.780         23.94           97.40      668.3          0.11720   
66        9.465         21.01           60.11      269.4          0.10440   
67       11.310         19.04           71.80      394.1          0.08139   
68        9.029         17.33           58.79      250.5          0.10660   
69       12.780         16.49           81.37      502.5          0.09831   

    mean compactness  mean concavity  mean concave points  mean symmetry  \
50           0.04966        0.016570             0.011150         0.1495   
51           0.06059        0.018570             0.017230         0.1353   
52           0.04751        0.019720             0.013490         0.1868   
53           0.14850        0.177200             0.106000         0.2092   
54           0.07081        0.052530             0.033340         0.1616   
55           0.05473        0.030360             0.022780         0.1920   
56           0.12670        0.132300             0.089940         0.1917   
57           0.13650        0.129300             0.081230         0.2027   
58           0.03789        0.000692             0.004167         0.1819   
59           0.05272        0.020610             0.007799         0.1683   
60           0.08061        0.010840             0.012900         0.2743   
61           0.08963        0.030000             0.009259         0.1828   
62           0.20080        0.213500             0.086530         0.1949   
63           0.08751        0.059880             0.021800         0.2341   
64           0.12620        0.112800             0.068730         0.1905   
65           0.14790        0.126700             0.090290         0.1953   
66           0.07773        0.021720             0.015040         0.1717   
67           0.04701        0.037090             0.022300         0.1516   
68           0.14130        0.313000             0.043750         0.2111   
69           0.05234        0.036530             0.028640         0.1590   

    mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
50                 0.05888  ...          25.72            82.98       516.5   
51                 0.05953  ...          23.19            96.08       656.7   
52                 0.06110  ...          21.33            83.67       527.2   
53                 0.06310  ...          24.13           135.10      1321.0   
54                 0.05684  ...          31.69           117.70      1030.0   
55                 0.05907  ...          22.47            81.81       506.2   
56                 0.05961  ...          28.14           170.10      2145.0   
57                 0.06758  ...          30.70           115.70       985.5   
58                 0.05501  ...          22.25            90.24       624.1   

In [ ]:
#EXTRACT FEATURES FOR TRAINING AND TESTING (x)
x=df.drop(['target'],axis = 'columns')
x=x.values

In [ ]:
#EXTRACT TARGET VARIABLES (y)
y=breast.target

In [ ]:
#SPLIT DATA INTO TRAINING(0.7) AND TESTING (0.3)
X_train, X_test, y_train, y_test = train_test_split(x, breast.target, test_size=0.3, random_state=100)

MODEL BUILDING

Naive Bayes and Logistic Regression

For the Naive Bayes, I am going to use the  Gaussian Because looking at the number of features and their nature, Gaussion will be best suitable for this task.

In [ ]:
#CREATE A NAIVE BAYES MODEL 
model = GaussianNB()
model.fit(X_train,y_train)
model.score(X_test,y_test)

0.9239766081871345

In [ ]:
model2=LogisticRegression(max_iter=1000000)
model2.fit(X_train,y_train)
model2.score(X_test,y_test)

0.9532163742690059

a) The model score for naive Bayes is **92%** and **95%** for the logistic Regression

In [ ]:
X2 = df[['mean texture','mean perimeter','mean smoothness','mean compactness','mean symmetry']]

In [ ]:
X2 = X2.values

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X2, breast.target, test_size=0.3, random_state=100)

In [ ]:
model = GaussianNB()
model.fit(X_train,y_train)
model.score(X_test,y_test)

0.9239766081871345

2. Logestic Regression

In [ ]:
model2=LogisticRegression(max_iter=1000000)
model2.fit(X_train,y_train)
model2.score(X_test,y_test)

0.8830409356725146

b) Using only 'mean texture','mean perimeter','mean smoothness','mean compactness',and 'mean symmetry', the model score for naive Bayes is **92%** and **88%** for the logistic Regression

Principal Component Analysis (PCA)

In [ ]:
#EXTRACT DATA FOR PCA
df_pca=df.drop(['target'],axis = 'columns')

#VIEW DATA SUMMARY
print(df_pca.iloc[:,0:4].describe())

       mean radius  mean texture  mean perimeter    mean area
count   569.000000    569.000000      569.000000   569.000000
mean     14.127292     19.289649       91.969033   654.889104
std       3.524049      4.301036       24.298981   351.914129
min       6.981000      9.710000       43.790000   143.500000
25%      11.700000     16.170000       75.170000   420.300000
50%      13.370000     18.840000       86.240000   551.100000
75%      15.780000     21.800000      104.100000   782.700000
max      28.110000     39.280000      188.500000  2501.000000


In [ ]:

#Scale independent variables, or x values
x = StandardScaler().fit_transform(df_pca)

#FORM A PANDAS DATAFRAME
x=pd.DataFrame(x,columns=breast.feature_names)

#VIEW SCALED DATA SUMMARY
print(x.iloc[:,0:4].describe())

        mean radius  mean texture  mean perimeter     mean area
count  5.690000e+02  5.690000e+02    5.690000e+02  5.690000e+02
mean  -3.153111e-15 -6.568462e-15   -6.993039e-16 -8.553985e-16
std    1.000880e+00  1.000880e+00    1.000880e+00  1.000880e+00
min   -2.029648e+00 -2.229249e+00   -1.984504e+00 -1.454443e+00
25%   -6.893853e-01 -7.259631e-01   -6.919555e-01 -6.671955e-01
50%   -2.150816e-01 -1.046362e-01   -2.359800e-01 -2.951869e-01
75%    4.693926e-01  5.841756e-01    4.996769e-01  3.635073e-01
max    3.971288e+00  4.651889e+00    3.976130e+00  5.250529e+00


PCA With n_components = 2

In [ ]:
# n_compents = 2
pca = PCA(n_components=2)

#transform and fit pca data
pca_x = pca.fit_transform(x)

#VIEW SHAPE
pca_x.shape

(569, 2)

In [ ]:
#VIEW PCA DATA
pca_x

array([[ 9.19283683,  1.94858307],
       [ 2.3878018 , -3.76817174],
       [ 5.73389628, -1.0751738 ],
       ...,
       [ 1.25617928, -1.90229671],
       [10.37479406,  1.67201011],
       [-5.4752433 , -0.67063679]])

In [ ]:
#PCA EXPLAINED RATIO
pca.explained_variance_ratio_

array([0.44272026, 0.18971182])

In [ ]:
#SPLIT PCA DATA INTO TRAINING(70%) AND TESTING(30%)
X_train_pca, X_test_pca, y_train, y_test = train_test_split(pca_x, y, test_size=0.3, random_state=30)

In [ ]:
#EVALUATE AND PREDICT WITH Naive Bayes
model = GaussianNB()
model.fit(X_train_pca, y_train)
model.score(X_test_pca,y_test)

0.9181286549707602

In [ ]:
#EVALUATE AND PREDICT WITH Logistic Regression
model2=LogisticRegression(max_iter=1000000)
model2.fit(X_train_pca,y_train)
model2.score( X_test_pca,y_test)

0.9473684210526315

c) Using 2 components for the PCA, Naive Bayes gave a model score of **92%** and logistic Regression attained a score of **95%**

PCA - Maintaining 95% information of the dataset

In [ ]:
#Maintaining 95% information of the dataset
pca = PCA(0.95)

pca_x = pca.fit_transform(x)

#VIEW SHAPE
pca_x.shape

(569, 10)

In [ ]:
pca.explained_variance_ratio_

array([0.44272026, 0.18971182, 0.09393163, 0.06602135, 0.05495768,
       0.04024522, 0.02250734, 0.01588724, 0.01389649, 0.01168978])

In [ ]:
pca.n_components_

10

d) i. The number of components to preserve 95% information of the dataset is  **10**

In [ ]:
#SPLIT PCA_95 DATA INTO TRAINING(70%) AND TESTING(30%)
X_train_pca, X_test_pca, y_train, y_test = train_test_split(pca_x, y, test_size=0.3, random_state=30)

In [ ]:
#EVALUATE AND PREDICT WITH Naive Bayes
model = GaussianNB()
model.fit(X_train_pca, y_train)
model.score(X_test_pca,y_test)

0.9181286549707602

In [ ]:
#EVALUATE AND PREDICT WITH Logistic Regression
model2=LogisticRegression(max_iter=1000000)
model2.fit(X_train_pca,y_train)
model2.score( X_test_pca,y_test)

0.9941520467836257

d) ii. After preserving 95% of components in the dataset, Naive Bayes attained a model score of **92%**, and Logistic regression attained an accuracy score of **99%**

e) Insights and Conclusion

From my observation, I learned that not all features are important to the target values, hence finding a way to elimate these variable will help in reducing training and computation time. For instance, the model score did not change significantly when some variables were eliminated. When I reduced the  components and scaled the features using PCA, Naive Bayes reduced small from aproximately 92% to 91.8%. While Logistics Regression score increases from 88% to close to 99%.

It can also be deduced that some features or variables may be relevant with some models and irrelevant to others. Therefore, it is good to check how variables relate and are important to the dependent varibale and which model is best for the task. 

My final recommendation for this task is that, if we will do Principal Component Analysis, Logistic Regression should be used, Otherwise Naive Bayes should be used if no PCA.